### PR TITLE
Make FST fully read-only and streamline FST constructor

### PR DESCRIPTION
**Pull request by dungba88**
### Description

This change will make the transition to off-heap FST writing easier.

- Streamline FST constructors so that they will all converge into a single constructor
- Move all write methods to FSTCompiler (finish and setEmptyOutput)

Eventually we would want to remove the constructor with `DataInput metaIn`

### CODE REVIEW

File: lucene/core/src/java/org/apache/lucene/util/fst/FST.java [hunk 1]
```diff
@@ -552,14 +527,11 @@ public void save(DataOutput metaOut, DataOutput out) throws IOException {
...
     metaOut.writeByte(t);
     metaOut.writeVLong(startNode);
-    if (bytes != null) {
-      long numBytes = bytes.getPosition();
-      metaOut.writeVLong(numBytes);
-      bytes.writeTo(out);
-    } else {
-      assert fstStore != null;
-      fstStore.writeTo(out);
+    // nocommit why does the FSTStore case does not need the numBytes?
```
Comment (dungba88) [inline lucene/core/src/java/org/apache/lucene/util/fst/FST.java:530]: I wasn't sure about this, need to debug on why the numBytes is not needed when writing with FSTStore

Comment (dungba88) [reply lucene/core/src/java/org/apache/lucene/util/fst/FST.java:530]: Found it, it is actually an issue: https://github.com/apache/lucene/issues/12697

Comment (mikemccand) [reply lucene/core/src/java/org/apache/lucene/util/fst/FST.java:530]: Egads, this looks like a serious bug?  What would a user need to do to tickle this?

Comment (dungba88) [reply lucene/core/src/java/org/apache/lucene/util/fst/FST.java:530]: Yeah, but it seems to be a rare case. User needs to use save a FSTStore-backed FST with a different DataOutput for metadata. I assume most people will use the same DataOutput for metadata and main FST body, which would not trigger the error.

Comment (dungba88) [reply lucene/core/src/java/org/apache/lucene/util/fst/FST.java:530]: Anyhow https://github.com/apache/lucene/pull/12698 should fix this, and we wouldn't need to handle 2 separate cases.

**Comment by mikemccand** [comment]
We closed this PR in favor of #12758?

**Comment by dungba88** [comment]
Yeah, this PR was originally opened for another purpose: consolidate the FSTStore and BytesStore, and that was already done.

**Pull request closed without merging**